src/oop/user_defined_classes/hard/q7/MainTest.java

```diff
@@ -0,0 +1,108 @@
+                missingFieldMessage);
+        assertTrue(classInstance.hasModifier(intAttributeName1, "private"),
+                incorrectModifierMessage);
+    }
+
+    @Test
+    public void courseClassHasRequiredConstructor() {
+        String missingConstructorMessage = "Your " + testClassName + " class is missing a required constructor.";
+        Class<?>[] arguments = {
+                String.class, int.class
+        };
+        assertTrue(classInstance.hasConstructor(arguments), missingConstructorMessage);
+    }
+
+    private static Stream<Arguments> constructorInputProvider() {
+        return Stream.of(
+                Arguments.of("Intro to Computer Science", 1),
+                Arguments.of("Editing and Publishing", 4),
+                Arguments.of("Advanced Theoretics of Quantum Mechanics and Their Application to Fish Cultivation", 9001),
+                Arguments.of("Master's Thesis", 50),
+                Arguments.of("Basic Life Skillz", 0),
+                Arguments.of("Still Life Painting", 10)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void courseConstructorInitializesFieldsCorrectly(String attribute1, int attribute2) throws Throwable {
+        Object[][] arguments = {
+                {attribute1, String.class},
+                {attribute2, int.class}
+        };
+        Object instance = classInstance.createInstance(arguments);
+        String incorrectFieldInstantiationMessage =
+                "Your " + testClassName + " constructor does not correctly initialize the object.";
+        _assertEquals(attribute1, classInstance.getFieldValue(instance, stringAttributeName1),
+                incorrectFieldInstantiationMessage);
+        _assertEquals(attribute2, classInstance.getFieldValue(instance, intAttributeName1),
+                incorrectFieldInstantiationMessage);
+    }
+
+    private static Stream<Arguments> methodInputProvider() {
+        return Stream.of(
+                Arguments.of("Artificial Intelligence 3", 4, 3, "Artificial Intelligence 3 is a difficult course"),
+                Arguments.of("Matrix Algebra", 2, 10, "Matrix Algebra is not a difficult course"),
+                Arguments.of("Renaissance History", 1, 1, "Renaissance History is a difficult course"),
+                Arguments.of("Recreating Jurassic Park for Fun and Profit", 800, 3, "Recreating Jurassic Park for Fun and Profit is a difficult course"),
+                Arguments.of("Clown Studies", 5, 1000, "Clown Studies is not a difficult course"),
+                Arguments.of("Nepotism", -1, 0, "Nepotism is not a difficult course"),
+                Arguments.of("Learn to Blink", -50, -25, "Learn to Blink is not a difficult course")
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("methodInputProvider")
+    public void correctDetermineDifficultyMethod(String attribute1, int attribute2, int dif, String mssg) throws Throwable {
```

Comment:
I don't think you actually need the String mssg here as a parameter you can get that with attribute 1 and a pre-defined message.

Reply:
yeah but it achieves the same end and I prefer doing it this way